# FIT1043 Introduction to Data Science

## Assignment 2

### Name:Ng Kai Yi

### Student id:32156944


## Introduction:
In this assignment,I will be reading a csv files which contain the features and labels.I will be analyzing which features I will include to train the model as this will have effect on the performance of the model.Since the analyzing technique is not taught in tutorial,so it is a good chance for me to do the research to improve myself.Then I will seperate the features and label.In this dataset,the label is 'score'.Then I will do the splitting of training and testing dataset.After that, I will normalize the datasets and use the training dataset to train and build my model.Then I will use the testing dataset to do the prediction of score.I will also construct the confusion matrix and analyze it.Then,I will also read a csv file which does not contain the score.I will predict the score for that dataset by using the model.Finally,I will also have a chance to get familiar with kaggle since I have to submit a csv file which contain the essayid and score predicted by using the model.

In [2]:
import pandas as pd #import pandas library to do data manipulation and analysis


In [3]:
dataset = pd.read_csv('FIT1043-Essay-Features.csv') 
#read the 'FIT1043-Essay-Features.csv' file into dataframe and dataset is assigned to it 
dataset # Display dataset

essayid  chars  words  commas  apostrophes  punctuations  \
0        1457   2153    426      14            6             0   
1         503   1480    292       9            7             0   
2         253   3964    849      19           26             1   
3         107    988    210       8            7             0   
4        1450   3139    600      13            8             0   
...       ...    ...    ...     ...          ...           ...   
1327     1151   2404    467      16           10             0   
1328     1015   1182    241       0           14             0   
1329     1345   1814    363       5           11             0   
1330      344   1427    287       5            8             0   
1331     1077   2806    542      24            6             0   

      avg_word_length  sentences  questions  avg_word_sentence         POS  \
0            5.053991         16          0          26.625000  423.995272   
1            5.068493         11          0          26.545455  290.993103   
2            4.669022         49          2          17.326531  843.990544   
3            4.704762         12          0          17.500000  207.653784   
4            5.231667         24          1          25.000000  594.652150   
...               ...        ...        ...                ...         ...   
1327         5.147752         22          0          21.227273  462.987069   
1328         4.904564         16          0          15.062500  238.655462   
1329         4.997245         13          3          27.923077  362.329640   
1330         4.972125         13          1          22.076923  284.657277   
1331         5.177122         22          3          24.636364  538.988889   

      POS/total_words  prompt_words  prompt_words/total_words  synonym_words  \
0            0.995294           207                  0.485915            105   
1            0.996552           148                  0.506849             77   
2            0.994100           285                  0.335689            130   
3            0.988828           112                  0.533333             62   
4            0.991087           255                  0.425000            165   
...               ...           ...                       ...            ...   
1327         0.991407           200                  0.428266            113   
1328         0.990272            94                  0.390041             67   
1329         0.998153           170                  0.468320            107   
1330         0.991837           144                  0.501742             83   
1331         0.994444           284                  0.523985            155   

      synonym_words/total_words  unstemmed  stemmed  score  
0                      0.246479        424      412      4  
1                      0.263699        356      345      4  
2                      0.153121        750      750      4  
3                      0.295238        217      209      3  
4                      0.275000        702      677      4  
...                         ...        ...      ...    ...  
1327                   0.241970        529      519      4  
1328                   0.278008        293      283      3  
1329                   0.294766        427      415      3  
1330                   0.289199        323      312      3  
1331                   0.285978        596      575      4  

[1332 rows x 19 columns]

In [4]:
dataset.describe() #generate descriptive analysis

essayid        chars        words       commas  apostrophes  \
count  1332.00000  1332.000000  1332.000000  1332.000000  1332.000000   
mean    905.27027  2101.745495   424.485736    14.667417     8.141141   
std     526.68760   865.963750   171.873730    10.920781     6.124520   
min       0.00000   169.000000    36.000000     0.000000     2.000000   
25%     442.75000  1527.250000   310.000000     7.000000     4.000000   
50%     914.50000  2029.500000   411.000000    13.000000     6.000000   
75%    1369.75000  2613.500000   525.000000    21.000000    11.000000   
max    1799.00000  6142.000000  1170.000000    72.000000    51.000000   

       punctuations  avg_word_length    sentences    questions  \
count    1332.00000      1332.000000  1332.000000  1332.000000   
mean        0.47973         4.939762    19.704204     1.222973   
std         1.27168         0.231071    19.202731     1.847446   
min         0.00000         2.231322     0.000000     0.000000   
25%         0.00000         4.791679    13.000000     0.000000   
50%         0.00000         4.946059    18.000000     1.000000   
75%         0.00000         5.092938    24.000000     2.000000   
max        26.00000         5.681429   642.000000    17.000000   

       avg_word_sentence          POS  POS/total_words  prompt_words  \
count        1332.000000  1332.000000      1332.000000   1332.000000   
mean           23.884687   420.596542         0.989935    198.913664   
std            11.160020   170.985111         0.007308     82.729266   
min             1.084112    35.647059         0.924771     14.000000   
25%            19.142857   305.406284         0.987758    144.000000   
50%            22.030331   406.982869         0.991572    193.000000   
75%            26.048234   520.739458         0.994425    246.000000   
max           303.000000  1158.984563         1.000000    669.000000   

       prompt_words/total_words  synonym_words  synonym_words/total_words  \
count               1332.000000     1332.00000                1332.000000   
mean                   0.469164      110.16967                   0.263846   
std                    0.052466       43.96192                   0.038870   
min                    0.288889       11.00000                   0.027299   
25%                    0.435709       81.00000                   0.238423   
50%                    0.465852      107.50000                   0.262872   
75%                    0.500000      134.00000                   0.288277   
max                    0.961207      355.00000                   0.465517   

         unstemmed      stemmed        score  
count  1332.000000  1332.000000  1332.000000  
mean    468.987988   455.507508     3.427177  
std     159.447449   155.751220     0.774275  
min      48.000000    50.000000     1.000000  
25%     361.000000   350.750000     3.000000  
50%     463.000000   448.000000     3.000000  
75%     581.000000   561.250000     4.000000  
max     750.000000   750.000000     6.000000

### Basic descriptive statistics of the values in the file:
There are total of 1332 essays with scores.The mean score of the essays is 3.427.The minimum score is 1 and the maximum score is 6.The average of the characters in the essay is 2101.746,and the maximum characters is 6142 while the minimum is 169,hence the range of the characters is great.The average words in the essay is 424.486 and the maximum words is 1170 whereas the minimum words is 36.There is essay which does not include any punctuations,questions,and commas,since the minimum in punctuations,questions and commas are 0.


In [5]:
X = dataset.iloc[:,[1,2,3,6,10,12,14,15,16,17]]
# X is assigned to the features that I wanted which include chars,words,commas,avg_word_length,POS,prompt_words,synonym_words,
# synonym_words / total_words, unstemmed and stemmed
X #Display X

chars  words  commas  avg_word_length         POS  prompt_words  \
0      2153    426      14         5.053991  423.995272           207   
1      1480    292       9         5.068493  290.993103           148   
2      3964    849      19         4.669022  843.990544           285   
3       988    210       8         4.704762  207.653784           112   
4      3139    600      13         5.231667  594.652150           255   
...     ...    ...     ...              ...         ...           ...   
1327   2404    467      16         5.147752  462.987069           200   
1328   1182    241       0         4.904564  238.655462            94   
1329   1814    363       5         4.997245  362.329640           170   
1330   1427    287       5         4.972125  284.657277           144   
1331   2806    542      24         5.177122  538.988889           284   

      synonym_words  synonym_words/total_words  unstemmed  stemmed  
0               105                   0.246479        424      412  
1                77                   0.263699        356      345  
2               130                   0.153121        750      750  
3                62                   0.295238        217      209  
4               165                   0.275000        702      677  
...             ...                        ...        ...      ...  
1327            113                   0.241970        529      519  
1328             67                   0.278008        293      283  
1329            107                   0.294766        427      415  
1330             83                   0.289199        323      312  
1331            155                   0.285978        596      575  

[1332 rows x 10 columns]

### Explanation on supervised machine learning and the notion of labelled data:
In supervised machine learning,we train the machine by using the labeled data and the supervised machine learning algorithm learn from the labelled input data and predict the output for unforeseen data.Labeled data means that the data is marked up and tagged with the correct answer we want the machine to predict and the data is attached with meaning.

### Explanation on training and test datasets:
Training dataset is a sample set of data used to train the model in order to predict the output for unforeseen data.Training dataset is also used to fit the parameters of classifier.

Test dataset is a sample set of data used to test the trained model,to evaluate the performance of the trained model.



In [5]:
y = dataset.iloc[:,-1].values
# y is assigned to the label,which is 'score'
y #Display y 

array([4, 4, 4, ..., 3, 3, 4], dtype=int64)

In [6]:
from sklearn.model_selection import train_test_split
# import train_test_split from sklearn.model_selection to split the data.

X_train, X_test, y_train, y_test = train_test_split(
 X, y, test_size = 0.25, random_state = 42
)
#Splitting features and label into training and testing datasets.X_train is assigned to features' training dataset and 
#X_test is assigned to features' testing dataset whereas y_train is assigned to the label's training dataset and 
#y_test is assigned to the label's testing dataset.


In [7]:
X_train #Display X_train

array([[2.09900000e+03, 4.20000000e+02, 1.10000000e+01, ...,
        2.97619048e-01, 4.47000000e+02, 4.42000000e+02],
       [1.18100000e+03, 2.43000000e+02, 0.00000000e+00, ...,
        3.00411523e-01, 2.80000000e+02, 2.66000000e+02],
       [1.10900000e+03, 2.31000000e+02, 1.30000000e+01, ...,
        2.07792208e-01, 2.89000000e+02, 2.84000000e+02],
       ...,
       [1.56300000e+03, 3.27000000e+02, 2.00000000e+00, ...,
        3.45565749e-01, 3.36000000e+02, 3.19000000e+02],
       [1.81700000e+03, 4.10000000e+02, 1.00000000e+01, ...,
        2.51219512e-01, 3.97000000e+02, 3.87000000e+02],
       [6.05000000e+02, 1.33000000e+02, 7.00000000e+00, ...,
        2.63157895e-01, 1.64000000e+02, 1.62000000e+02]])

In [8]:
X_test #Display X_test

array([[2.36900000e+03, 4.87000000e+02, 7.00000000e+00, ...,
        2.62833676e-01, 4.68000000e+02, 4.56000000e+02],
       [2.61900000e+03, 5.12000000e+02, 8.00000000e+00, ...,
        2.34375000e-01, 6.49000000e+02, 6.27000000e+02],
       [1.45900000e+03, 3.13000000e+02, 1.90000000e+01, ...,
        2.93929712e-01, 3.80000000e+02, 3.66000000e+02],
       ...,
       [1.97000000e+03, 3.87000000e+02, 1.10000000e+01, ...,
        2.53229974e-01, 4.38000000e+02, 4.16000000e+02],
       [1.15600000e+03, 2.19000000e+02, 1.00000000e+01, ...,
        2.37442922e-01, 2.97000000e+02, 2.95000000e+02],
       [2.30100000e+03, 4.98000000e+02, 1.70000000e+01, ...,
        2.57028112e-01, 5.49000000e+02, 5.34000000e+02]])

In [9]:
y_train #Display y_train

array([3, 3, 3, 3, 3, 4, 4, 3, 6, 2, 3, 2, 4, 4, 3, 4, 4, 4, 5, 3, 2, 3,
       4, 4, 4, 2, 4, 4, 4, 1, 3, 5, 2, 3, 4, 1, 4, 4, 3, 4, 3, 3, 4, 4,
       3, 3, 3, 5, 2, 3, 3, 4, 3, 3, 3, 4, 2, 2, 3, 3, 3, 4, 2, 3, 3, 4,
       4, 3, 4, 4, 3, 3, 2, 4, 3, 4, 3, 4, 3, 4, 3, 4, 4, 2, 4, 4, 4, 2,
       5, 3, 4, 3, 4, 5, 4, 4, 3, 3, 4, 4, 2, 3, 4, 2, 3, 4, 4, 4, 3, 3,
       3, 4, 4, 4, 3, 4, 4, 3, 5, 4, 4, 4, 3, 3, 3, 4, 4, 3, 3, 4, 3, 3,
       4, 1, 3, 3, 3, 3, 3, 4, 3, 4, 4, 2, 4, 3, 3, 3, 4, 4, 2, 4, 4, 4,
       5, 3, 4, 3, 2, 4, 3, 4, 4, 3, 4, 3, 4, 2, 3, 3, 5, 3, 3, 3, 4, 4,
       3, 3, 4, 3, 2, 5, 3, 4, 3, 3, 4, 3, 3, 3, 4, 3, 4, 1, 2, 4, 4, 4,
       4, 2, 5, 3, 4, 2, 4, 3, 4, 3, 4, 4, 3, 4, 3, 4, 3, 4, 3, 3, 3, 3,
       3, 4, 3, 4, 4, 4, 4, 3, 3, 3, 4, 3, 3, 2, 3, 3, 4, 3, 4, 3, 4, 3,
       4, 3, 3, 4, 4, 4, 5, 3, 4, 3, 4, 3, 4, 2, 4, 4, 4, 3, 3, 4, 3, 3,
       3, 2, 3, 3, 3, 3, 4, 3, 4, 4, 2, 3, 3, 4, 3, 4, 2, 5, 4, 4, 4, 3,
       3, 4, 3, 4, 3, 3, 2, 5, 2, 5, 3, 4, 4, 4, 3,

In [10]:
y_test #Display y_test

array([3, 4, 3, 2, 3, 4, 4, 4, 2, 4, 4, 4, 3, 4, 4, 4, 4, 3, 3, 3, 4, 4,
       2, 4, 2, 3, 4, 3, 3, 3, 3, 5, 3, 5, 3, 3, 3, 3, 4, 3, 2, 4, 4, 4,
       4, 3, 4, 1, 4, 3, 4, 3, 4, 4, 4, 2, 4, 3, 4, 4, 4, 3, 6, 4, 4, 5,
       3, 4, 3, 3, 3, 4, 3, 2, 4, 2, 4, 4, 3, 5, 3, 2, 5, 4, 4, 4, 4, 4,
       1, 3, 3, 3, 4, 4, 3, 4, 4, 4, 3, 4, 3, 3, 4, 4, 4, 5, 3, 4, 4, 3,
       4, 3, 3, 3, 4, 5, 3, 3, 4, 3, 3, 2, 4, 4, 3, 4, 4, 2, 3, 4, 4, 3,
       4, 3, 3, 3, 2, 3, 3, 4, 3, 4, 4, 3, 3, 4, 4, 3, 4, 3, 3, 4, 4, 5,
       3, 1, 3, 4, 1, 3, 4, 3, 3, 2, 3, 4, 4, 3, 4, 3, 3, 3, 5, 3, 4, 3,
       3, 4, 3, 3, 3, 3, 2, 4, 3, 4, 3, 3, 3, 2, 3, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 5, 4, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 3, 4, 3, 3, 3, 2, 4,
       2, 3, 3, 3, 5, 3, 3, 3, 4, 2, 3, 4, 3, 4, 4, 3, 4, 4, 4, 3, 4, 4,
       2, 4, 3, 4, 4, 3, 3, 3, 3, 4, 4, 4, 4, 4, 4, 1, 3, 3, 4, 4, 3, 4,
       4, 4, 4, 3, 4, 4, 4, 4, 5, 4, 2, 4, 4, 4, 5, 5, 4, 4, 4, 4, 4, 3,
       4, 4, 2, 3, 3, 3, 4, 2, 3, 4, 3, 3, 4, 4, 3,

### Difference between binary and multi-class classification:
Binary classification is a type of supervised learning because the training dataset is labelled.Moreover,Binary classification is the classification that involve only two class labels.

Binary classification is the task of classifying given dataset into one of two classes based on the features,one will be the normal state and the other one will be abnormal state.For example,medical testing on whether the patient has breast cancer,'the patient has breast cancer'class will be the abnormal state whereas 'the patient does not have breast cancer'is the normal state.

The algorithm commonly used for the binary classification is logistic regression,decision trees,random forests,support vector machine and neural networks.The results of the binary classification can be represented by the (2x2) matrix.

### --------------------------------------------------------------------------------------------------------------------------------------------------------------

Multi-class classification is also a type of supervised learning as the training dataset is labelled.Multi-class classification 
is the classification that involve more than two class labels.

Multi-class classification is the task of classifying the samples into one of three or more classes based on the features.Each sample can only be assigned to one label out of many.For instance,classification of the species of dog based on the features on a given image,the dog can only be labelled as 'poodle' but not 'poodle' and 'bulldog'.

The algorithm commonly used for the multi-class classification is the k-Nearest Neighbors,decision trees,random forests and naive bayes.There are ways to reduce the problem of multi-class classification to multiple binary classification like one vs rest and one vs one.The results of the binary classification can be represented by the (3x3) or larger dimension matrix.

### Explanation on the purpose of normalised /scaling of data:
Most of the machine learning algorithms calculate the distance between two data points by Eucledian distance to determine their similarities.Hence if one the features has a great range of values then the distance calculated will be greatly affected by this feature.Moreover,if the features have different scale,the feature with greater magnitude will tend to be given greater weightage.Therefore,we normalise/scale the data before we apply the distance based algorithm so that all the features contribute proportionally to the final distance.

In [11]:
from sklearn.preprocessing import StandardScaler
#importing StandardScaler from sklearn.preprocessing
sc = StandardScaler()
X_train = sc.fit_transform(X_train) # Scale training dataset so that it has the mean of 0 and certain standard deviation
X_test = sc.transform(X_test) 
# normalize the testing dataset by using the mean of 0 and certain standard deviation stored in the StandardScaler()


In [12]:
X_train # Display features' training dataset

array([[-6.94014264e-04, -2.35692855e-02, -3.35382185e-01, ...,
         8.59185435e-01, -1.40768375e-01, -8.90135489e-02],
       [-1.07399585e+00, -1.06520682e+00, -1.32160505e+00, ...,
         9.31528414e-01, -1.19262533e+00, -1.22246649e+00],
       [-1.15817639e+00, -1.13582631e+00, -1.56068938e-01, ...,
        -1.46790475e+00, -1.13593843e+00, -1.10654517e+00],
       ...,
       [-6.27371340e-01, -5.70870363e-01, -1.14229180e+00, ...,
         2.10131193e+00, -8.39906830e-01, -8.81142594e-01],
       [-3.30401115e-01, -8.24188638e-02, -4.25038809e-01, ...,
        -3.42859577e-01, -4.55695607e-01, -4.43217593e-01],
       [-1.74744014e+00, -1.71255218e+00, -6.94008681e-01, ...,
        -3.35790329e-02, -1.92325651e+00, -1.89223414e+00]])

In [13]:
X_test #Display features' testing dataset 

array([[ 3.14982997e-01,  3.70722889e-01, -6.94008681e-01, ...,
        -4.19783805e-02, -8.49893712e-03,  1.14748057e-03],
       [ 6.07276525e-01,  5.17846834e-01, -6.04352057e-01, ...,
        -7.79240269e-01,  1.13153764e+00,  1.10240006e+00],
       [-7.48965448e-01, -6.53259773e-01,  3.81870805e-01, ...,
         7.63608015e-01, -5.62770866e-01, -5.78459138e-01],
       ...,
       [-1.51517475e-01, -2.17772894e-01, -3.35382185e-01, ...,
        -2.90775741e-01, -1.97455276e-01, -2.56455461e-01],
       [-1.10322520e+00, -1.20644581e+00, -4.25038809e-01, ...,
        -6.99761450e-01, -1.08555007e+00, -1.03570436e+00],
       [ 2.35479157e-01,  4.35457425e-01,  2.02557558e-01, ...,
        -1.92379653e-01,  5.01683179e-01,  5.03473216e-01]])

### Explanation on SVM:
Support vector machine find the best hyperplane which seperate the classes by finding the greatest margin,which is the distance between the line and the support vector while linear regression find the best line by using the equation Y = a + bX, X is the explanatory variable,Y is the dependent variable and a is the intercept and b is the slope.Support vector machine is based on the geometrical properties of the data whereas linear regression is based on the statistical way.

### Explanation on kernel in SVM:
Kernel in Support Vector Machine is used to refer the kernel trick,which is a set of mathematical functions used by the Support Vector Machine.It transform the data which cannot be seperated linearly to the data which can be seperated linearly.Kernel function take the data as input and transform it to a high-dimensional feature space so that it can be seperated linearly.There are different types of kernel functions which are 'linear','non-linear','polynomial','radial basis function' and 'sigmoid'.

In [14]:
from sklearn import svm # importing svm from sklearn 
model = svm.SVC(C = 1000, gamma = 0.001, kernel = 'rbf') 
# Use Support Vector Classifier as classifer and C is set to 1000, gamma is set to 0.001 and kernel is radial basis function
# and model is assigned to it.
model.fit(X_train, y_train)
# Fit the features' training data and label's training data to the model.


SVC(C=1000, gamma=0.001)

In [15]:
y_pred = model.predict(X_test) 
# Predict the score by fitting features' testing dataset y_pred is assigned to it
y_pred
# Display the predicted score.

array([4, 4, 3, 1, 3, 4, 3, 4, 2, 4, 4, 4, 3, 3, 4, 3, 3, 3, 3, 3, 4, 3,
       2, 4, 3, 4, 4, 3, 3, 4, 4, 4, 3, 4, 3, 3, 4, 3, 4, 3, 2, 4, 4, 4,
       4, 3, 3, 2, 4, 2, 4, 3, 4, 4, 4, 2, 4, 3, 4, 4, 4, 3, 5, 4, 4, 4,
       3, 4, 4, 3, 3, 4, 3, 2, 4, 3, 4, 4, 3, 4, 3, 3, 4, 4, 3, 3, 4, 4,
       2, 3, 4, 3, 4, 4, 3, 4, 4, 4, 3, 3, 4, 3, 3, 4, 4, 4, 4, 4, 4, 3,
       4, 4, 4, 3, 4, 4, 3, 4, 4, 4, 3, 2, 4, 3, 3, 4, 4, 3, 4, 3, 3, 4,
       4, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 4, 3, 4, 4, 3, 4, 4, 3, 3, 4, 3,
       3, 2, 4, 4, 2, 3, 4, 3, 4, 2, 3, 4, 4, 3, 4, 3, 4, 3, 4, 3, 4, 4,
       3, 4, 3, 3, 3, 3, 2, 4, 3, 4, 3, 3, 3, 2, 3, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 3, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 4, 3, 4, 4, 2, 3,
       3, 3, 4, 3, 4, 3, 3, 3, 5, 2, 3, 4, 4, 4, 4, 3, 4, 4, 3, 3, 4, 4,
       2, 4, 3, 4, 4, 3, 2, 3, 3, 3, 4, 3, 4, 4, 3, 2, 3, 3, 4, 3, 3, 4,
       3, 4, 4, 4, 4, 3, 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 4,
       3, 4, 2, 3, 3, 3, 3, 2, 3, 4, 3, 4, 4, 4, 3,

In [16]:
from sklearn.metrics import confusion_matrix
#Importing confusion_matrix from sklearn.metrics
cm = confusion_matrix(y_test, y_pred) 
#Construct confusion matrix by using label's testing dataset and predicted score.
cm
#Display confusion matrix

array([[  0,   5,   0,   0,   0,   0],
       [  1,  15,   8,   0,   0,   0],
       [  0,   4,  94,  35,   0,   0],
       [  0,   0,  30, 125,   1,   0],
       [  0,   0,   1,  13,   0,   0],
       [  0,   0,   0,   0,   1,   0]], dtype=int64)

### Explanation on confusion matrix:
Confusion matrix is a N x N matrix which is used to assess the performance of the classification model,N is the number of target classes.Confusion matrix will compare the predicted value with the actual values,hence we can see what errors our model has made and how well it perform.

The confusion matrix I constructed based on y_test and y_pred showed that for score 1, 5 and 6,the model failed to classify any of the samples to their actual score,which means there is no True Positive for score 1, 5 and 6.Therefore,the precision for score 1, 5 and 6 is 0.For score 2, 15 out of 24 samples correctly classified to their actual score,the precision is 0.62.For score 3,94 out of 133 samples correctly classified to their actual score,the precision is 0.71,for score 4, 124 out of 156 correctly classified to their actual score,the precision is 0.72.The accuracy is calculated as the ratio of the total number of samples correctly classified to the total number of samples.The accuracy is 0.703.

In [17]:
from sklearn.metrics import classification_report
#Importing classification_report from sklearn.metrics
print(classification_report(y_test, y_pred))
#Printing the classification_report based on the label's testing dataset and predicted score.

              precision    recall  f1-score   support

           1       0.00      0.00      0.00         5
           2       0.62      0.62      0.62        24
           3       0.71      0.71      0.71       133
           4       0.72      0.80      0.76       156
           5       0.00      0.00      0.00        14
           6       0.00      0.00      0.00         1

    accuracy                           0.70       333
   macro avg       0.34      0.36      0.35       333
weighted avg       0.67      0.70      0.68       333



C:\Users\User\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:
from sklearn import metrics #importing metrics from sklearn
metrics.accuracy_score(y_test, y_pred) 
#Get the accuracy of the confusion matrix constructed based on label's testing dataset and predicted score

0.7027027027027027

### Explanation on QWK:
QWK stands for quadratic weighted kappa.Quadratic weighted kappa is the measure of agreement between the predicted data by the model and the labelled data of the same object.Firstly,it will start with the accuracy by finding the proportion of all the objects which is classfied into the actual class(label).Then the value will be adjusted based on the probability of the model's predictions are same as the label(actual class) by chance.QWK can range from -1 to 1, 0 means random agreement between a model's prediction and the label(actual class), 1 means perfect agreement, and if Quadratic weighted kappa falls below 0,it indicates less than chance agreement.If it is between 0.01 to 0.20,it means poor agreement,if it is between 0.21 to 0.40,it means fair agreement,if it is between 0.41 to 0.60,it means moderate agreement,if it is between 0.61 to 0.80,it means substantial agreement,if it is between 0.81 to 1.00,it means almost perfect agreement

### --------------------------------------------------------------------------------------------------------------------------------------------------------------
This explanation is based on this assignment,which the label is score,which has 6 possible score,range from 1-6.
To calculate Quadratic Weighted Kappa,

1)Firstly,we have to construct a 6 x 6 confusion matrix,(O) based on the predicted score and the actual score.

2)Then we construct a 6 x 6 matrix of weights,(w) based on the difference between the actual and the predicted score.

3)Construct the actual score's histogram vector and the predicted score's histogram vector.

4)Then construct a 6 x 6 matrix of expected score,(E),which is calculated by the outer product of actual score's histogram vector and the predicted score's histogram vectoconfusion matrixr.

5)Then normalize matrix of expected score and confusion matrix so that they have the same sum.

6)Then the quadratic weighted kappa can be calculated by using 3 matrices constructed.
  Quadratic weighted kappa = 1 -  [(∑i,j wi,j Oij) / (∑i,j wi,j Ei,j.)]

In [19]:
from sklearn.metrics import cohen_kappa_score
#importing cohen_kappa_score from sklearn.metrics
cohen_kappa_score(y_test,y_pred, labels=None, weights='quadratic', sample_weight=None) 
#calculate the Quadratic weighted kappa score by using the function cohen_kappa_score and fitting in y_test and y_pred.


0.6957433086101257

### Explanation on QWK score:
The QWK score I got is 0.6957. 0.6957 is between 0.61 and 0.80 which indicates that the agreement between the score predicted by the model and the actual score is substantial.

In [20]:
data = pd.read_csv('FIT1043-Essay-Features-Submission.csv')
#read the 'FIT1043-Essay-Features-Submission.csv' file into dataframe and data is assigned to it 
data #Display data

essayid  chars  words  commas  apostrophes  punctuations  \
0       1623   4332    900      28           13             0   
1       1143   1465    280      11            3             1   
2        660   1696    325      17            2             0   
3       1596   2640    555      20           17             0   
4        846   2844    596      33            4             1   
..       ...    ...    ...     ...          ...           ...   
194     1226   1208    242       8            8             0   
195      862   4039    817      24           11             1   
196     1562   2448    468      22            7             0   
197     1336   1081    214      14            5             0   
198     1171   2094    433      11           12             0   

     avg_word_length  sentences  questions  avg_word_sentence         POS  \
0           4.813333         39          1          23.076923  893.988852   
1           5.232143         14          3          20.000000  278.321343   
2           5.218462         19          1          17.105263  321.316770   
3           4.756757         28          0          19.821429  551.989150   
4           4.771812         24          9          24.833333  593.658810   
..               ...        ...        ...                ...         ...   
194         4.991736         13          0          18.615385  237.327684   
195         4.943696         47          2          17.382979  812.656033   
196         5.230769         22          0          21.272727  465.656652   
197         5.051402         11          0          19.454545  212.990566   
198         4.836028         19          0          22.789474  426.651090   

     POS/total_words  prompt_words  prompt_words/total_words  synonym_words  \
0           0.993321           392                  0.435556            196   
1           0.994005           131                  0.467857             51   
2           0.988667           178                  0.547692             92   
3           0.994575           228                  0.410811            107   
4           0.996072           279                  0.468121            138   
..               ...           ...                       ...            ...   
194         0.980693           135                  0.557851             58   
195         0.994683           386                  0.472460            210   
196         0.994993           224                  0.478632            101   
197         0.995283           114                  0.532710             63   
198         0.985337           221                  0.510393            121   

     synonym_words/total_words  unstemmed  stemmed  
0                     0.217778        750      750  
1                     0.182143        339      316  
2                     0.283077        352      337  
3                     0.192793        632      605  
4                     0.231544        626      607  
..                         ...        ...      ...  
194                   0.239669        244      242  
195                   0.257038        750      750  
196                   0.215812        540      526  
197                   0.294393        259      256  
198                   0.279446        501      478  

[199 rows x 18 columns]

In [21]:
x_value = data.iloc[:,[1,2,3,6,10,12,14,15,16,17]].values 
# x_value is assigned to the features that I wanted which include chars,words,commas,avg_word_length,POS,prompt_words,
#synonym_words,synonym_words / total_words, unstemmed and stemmed
x_value
# Display x_value

array([[4.33200000e+03, 9.00000000e+02, 2.80000000e+01, ...,
        2.17777778e-01, 7.50000000e+02, 7.50000000e+02],
       [1.46500000e+03, 2.80000000e+02, 1.10000000e+01, ...,
        1.82142857e-01, 3.39000000e+02, 3.16000000e+02],
       [1.69600000e+03, 3.25000000e+02, 1.70000000e+01, ...,
        2.83076923e-01, 3.52000000e+02, 3.37000000e+02],
       ...,
       [2.44800000e+03, 4.68000000e+02, 2.20000000e+01, ...,
        2.15811966e-01, 5.40000000e+02, 5.26000000e+02],
       [1.08100000e+03, 2.14000000e+02, 1.40000000e+01, ...,
        2.94392523e-01, 2.59000000e+02, 2.56000000e+02],
       [2.09400000e+03, 4.33000000e+02, 1.10000000e+01, ...,
        2.79445727e-01, 5.01000000e+02, 4.78000000e+02]])

In [22]:
x_value = sc.transform(x_value) #Normalize x_value
x_value #Display x_value

array([[ 2.61007178,  2.80121047,  1.18878042, ..., -1.20921457,
         1.76769065,  1.8945291 ],
       [-0.7419504 , -0.84746338, -0.33538219, ..., -2.13238714,
        -0.8210112 , -0.90046281],
       [-0.47187118, -0.58264028,  0.20255756, ...,  0.48245131,
        -0.73913012, -0.76522127],
       ...,
       [ 0.40734775,  0.25890869,  0.65084068, ..., -1.26014168,
         0.44499628,  0.45195263],
       [-1.19091326, -1.2358706 , -0.06641231, ...,  0.77559778,
        -1.32489477, -1.28686723],
       [-0.00653988,  0.05293517, -0.33538219, ...,  0.38838008,
         0.19935304,  0.1428291 ]])

In [23]:
y_predicted = model.predict(x_value) 
# Predict the score by fitting x_value into the model and y_predicted is assigned to it.
y_predicted #Display y_predicted

array([4, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 3,
       4, 4, 4, 4, 4, 4, 4, 3, 2, 4, 3, 3, 4, 3, 4, 4, 3, 3, 3, 4, 3, 3,
       2, 3, 3, 4, 4, 3, 3, 4, 4, 4, 3, 4, 3, 4, 4, 4, 2, 3, 3, 4, 3, 4,
       4, 4, 3, 4, 3, 4, 4, 3, 3, 3, 3, 4, 4, 4, 4, 3, 4, 3, 4, 2, 4, 4,
       2, 3, 3, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 4, 2, 3, 3, 4, 3, 3, 4, 4,
       5, 3, 4, 4, 4, 3, 3, 4, 2, 3, 4, 4, 4, 3, 2, 4, 3, 4, 3, 2, 4, 4,
       2, 3, 4, 3, 3, 4, 2, 3, 4, 4, 4, 3, 3, 4, 4, 3, 4, 4, 4, 3, 4, 3,
       4, 3, 3, 3, 4, 4, 4, 3, 2, 3, 4, 4, 3, 3, 2, 4, 3, 4, 3, 3, 4, 4,
       4, 3, 3, 4, 4, 3, 4, 4, 4, 4, 3, 3, 4, 3, 3, 4, 3, 4, 3, 4, 4, 3,
       3], dtype=int64)

In [24]:
res = pd.DataFrame(y_predicted) # Transform y_predicted into dataframe and res is assigned to it
res #Display res

0
0    4
1    3
2    3
3    4
4    4
..  ..
194  3
195  4
196  4
197  3
198  3

[199 rows x 1 columns]

In [25]:
res.columns = ['score']  # res now has the heading 'score'
res #Display res

score
0        4
1        3
2        3
3        4
4        4
..     ...
194      3
195      4
196      4
197      3
198      3

[199 rows x 1 columns]

In [26]:
essayid = data.iloc[:,[0]] # essayid is assigned to the first column of data which is 'essayid'
essayid #Display essayid

essayid
0       1623
1       1143
2        660
3       1596
4        846
..       ...
194     1226
195      862
196     1562
197     1336
198     1171

[199 rows x 1 columns]

In [27]:
result = essayid.join(res) # Joining essayid and res on index

In [28]:
result.to_csv('32156944-Ng Kai Yi-1.csv', index = None) # Writing the result to csv file '32156944-Ng Kai Yi-1.csv'

### Conclusion:
After completing this assignment,I have more knowledge on how to build a model with better performance.I did some research on how to tune the hyperparameter of SVC so that the predicted data can have higher accuracy and better quadratic weighted kappa score.Moreover,I also gained knowledge on how to calculate quadratic weighted kappa score and what it indicates.Not to mention,I also learnt how confusion matrix work and how to analyze the confusion matrix by calculating accuracy,precision and recall.Furthermore,I also learnt how to differentiate binary classification and multi-class classification.I found that it is really interesting to compete among my friends on kaggle eventhough it is very challenging when I try to improve my QWK score hence it is a really good experience to me.